### PR TITLE
Introduced indexes builder

### PR DESCRIPTION
**Pull request by ahmetmircik**
addressed https://github.com/hazelcast/hazelcast/pull/13359#discussion_r209520150 and https://github.com/hazelcast/hazelcast/pull/13359#discussion_r209520049

- Introduced `Indexes#Builder`
- Moved `IndexProvider` and `DefaultIndexProvider` to `com.hazelcast.query.impl`

ee counterpart: https://github.com/hazelcast/hazelcast-enterprise/pull/2314

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/query/impl/Indexes.java [hunk 1]
```diff
@@ -270,20 +243,52 @@ public IndexesStats getIndexesStats() {
...
 
-    private QueryContextProvider createQueryContextProvider(MapConfig mapConfig, boolean global) {
-        if (mapConfig.isStatisticsEnabled()) {
-            return global ? new GlobalQueryContextProviderWithStats() : new PartitionQueryContextProviderWithStats(this);
-        } else {
-            return global ? new GlobalQueryContextProvider() : new PartitionQueryContextProvider(this);
+    /**
+     * Builder which is used to create a new Indexes object.
+     */
+    public static final class Builder {
```
Comment (taburet) [inline hazelcast/src/main/java/com/hazelcast/query/impl/Indexes.java]: Javadocs on methods would be useful.

File: hazelcast/src/main/java/com/hazelcast/query/impl/Indexes.java [hunk 2]
```diff
@@ -16,109 +16,80 @@
...
-     *
-     * @param serializationService the serialization service to be used by the
-     *                             new indexes.
-     * @param copyBehavior         the desired copy behaviour of the new indexes.
-     * @return the constructed indexes.
-     */
-    public static Indexes createStandaloneIndexes(InternalSerializationService serializationService,
-                                                  IndexCopyBehavior copyBehavior) {
-        return new Indexes(serializationService, copyBehavior);
+    public static Builder newBuilder(IndexCopyBehavior indexCopyBehavior, SerializationService ss) {
```
Comment (taburet) [inline hazelcast/src/main/java/com/hazelcast/query/impl/Indexes.java]: Missing javadoc.

File: hazelcast/src/main/java/com/hazelcast/query/impl/Indexes.java (hunk 1 above)
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/query/impl/Indexes.java]: fixed

File: hazelcast/src/main/java/com/hazelcast/query/impl/Indexes.java (hunk 2 above)
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/query/impl/Indexes.java]: fixed

**Pull request merged**